main/src/ca/uwaterloo/flix/language/phase/unification/ContextReduction.scala

```diff
@@ -0,0 +1,115 @@
+  def entail(instances: MultiMap[Symbol.ClassSym, ResolvedAst.Instance], tconstrs0: List[TypedAst.TypeConstraint], tconstr: TypedAst.TypeConstraint)(implicit flix: Flix): Boolean = {
+    // tconstrs0 entail tconstr if tconstr is a superclass of any member or tconstrs0
+    // or if there is an instance matching tconstr and all of the instance's constraints are entailed by tconstrs0
+    tconstrs0.exists(bySuper(instances, _).contains(tconstr)) || {
+      byInst(instances, tconstr) match {
+        case Result.Ok(tconstrs) => tconstrs.forall(entail(instances, tconstrs0, _))
+        case Result.Err(_) => false
+      }
+    }
+  }
+
+  /**
+    * Removes the type constraints which are entailed by the others in the list.
+    */
+  private def simplify(instances: MultiMap[Symbol.ClassSym, ResolvedAst.Instance], tconstrs0: List[TypedAst.TypeConstraint])(implicit flix: Flix): List[TypedAst.TypeConstraint] = {
+
+    @tailrec
+    def loop(acc: List[TypedAst.TypeConstraint], tconstrs0: List[TypedAst.TypeConstraint]): List[TypedAst.TypeConstraint] = tconstrs0 match {
+      case Nil => acc
+      case head :: tail if entail(instances, acc ++ tail, head) => loop(acc, tail)
+      case head :: tail => loop(head :: acc, tail)
+    }
+
+    loop(Nil, tconstrs0)
+  }
+
+  /**
+    * Normalizes a list of type constraints, converting to head-normal form and removing semantic duplicates.
+    */
+  private def reduce(instances: MultiMap[Symbol.ClassSym, ResolvedAst.Instance], tconstrs0: List[TypedAst.TypeConstraint])(implicit flix: Flix): Result[List[TypedAst.TypeConstraint], UnificationError] = {
+    for {
+      tconstrs <- Result.sequence(tconstrs0.map(toHeadNormalForm(instances, _)))
+    } yield simplify(instances, tconstrs.flatten)
+  }
+
+  /**
+    * Converts the type constraint to head-normal form, i.e. `a[X1, Xn]`, where `a` is a variable and `n >= 0`.
+    */
+  private def toHeadNormalForm(instances: MultiMap[Symbol.ClassSym, ResolvedAst.Instance], tconstr: TypedAst.TypeConstraint)(implicit flix: Flix): Result[List[TypedAst.TypeConstraint], UnificationError] = {
+    if (isHeadNormalForm(tconstr.arg)) {
+      List(tconstr).toOk
+    } else {
+      byInst(instances, tconstr)
+    }
+  }
+
+  /**
+    * Returns the list of constraints that hold if the given constraint `tconstr` holds, using the constraints on available instances.
+    */
+  private def byInst(instances0: MultiMap[Symbol.ClassSym, ResolvedAst.Instance], tconstr: TypedAst.TypeConstraint)(implicit flix: Flix): Result[List[TypedAst.TypeConstraint], UnificationError] = {
+    val matchingInstances = instances0(tconstr.sym).toList
+
+    def tryInst(inst: ResolvedAst.Instance): Result[List[TypedAst.TypeConstraint], UnificationError] = {
+      for {
+        subst <- Unification.unifyTypes(tconstr.arg, inst.tpe)
+      } yield inst.tconstrs.map(subst(_))
+    }
+
+    matchingInstances.map(tryInst).find {
```

Comment:
This is unique because of prior checks?

Reply:
Instead of using `find` this way, are you familiar with Scala's `collect`?

Reply:
We could also find *all* instances that match. Check if there is 1, return it, and otherwise throw a compiler exception.

That is less fast, but I think it might be more clean at this point. It would allow us to understand if we have made a mistake somewhere else. If you agree with this, I would just map/filter/collect over instances and check the size of the result.